transports/quic/src/transport.rs

```diff
@@ -0,0 +1,641 @@
+    type Error = TransportError;
+    type ListenerUpgrade = Upgrade;
+    type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn listen_on(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<ListenerId, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or(CoreTransportError::MultiaddrNotSupported(addr))?;
+        let listener_id = ListenerId::new();
+        let listener = Listener::new::<P>(listener_id, socket_addr, self.config.clone())
+            .map_err(CoreTransportError::Other)?;
+        self.listeners.push(listener);
+        // Drop reference to dialer endpoint so that the endpoint is dropped once the last
+        // connection that uses it is closed.
+        // New outbound connections will use a bidirectional (listener) endpoint.
+        match socket_addr {
+            SocketAddr::V4(_) => self.ipv4_dialer.take(),
+            SocketAddr::V6(_) => self.ipv6_dialer.take(),
+        };
+        Ok(listener_id)
+    }
+
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        if let Some(listener) = self.listeners.iter_mut().find(|l| l.listener_id == id) {
+            listener.close(Ok(()));
+            true
+        } else {
+            false
+        }
+    }
+
+    fn address_translation(&self, _server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        Some(observed.clone())
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| CoreTransportError::MultiaddrNotSupported(addr.clone()))?;
+        if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
+            return Err(CoreTransportError::MultiaddrNotSupported(addr));
+        }
+        let mut listeners = self
+            .listeners
+            .iter_mut()
+            .filter(|l| {
+                let listen_addr = l.endpoint_channel.socket_addr();
+                listen_addr.is_ipv4() == socket_addr.is_ipv4()
+                    && listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()
+            })
+            .collect::<Vec<_>>();
+
+        let (tx, rx) = oneshot::channel();
+        let to_endpoint = ToEndpoint::Dial {
+            addr: socket_addr,
+            result: tx,
+        };
+        if listeners.is_empty() {
+            let dialer = match socket_addr {
+                SocketAddr::V4(_) => &mut self.ipv4_dialer,
+                SocketAddr::V6(_) => &mut self.ipv6_dialer,
+            };
+            if dialer.is_none() {
+                let _ = dialer.insert(Dialer::new::<P>(
+                    self.config.clone(),
+                    socket_addr.is_ipv6(),
+                )?);
+            }
+            dialer
+                .as_mut()
+                .unwrap()
+                .pending_dials
+                .push_back(to_endpoint);
```

Comment:
This can be expressed a bit simpler IMO:

```suggestion
            let maybe_dialer = match socket_addr {
                SocketAddr::V4(_) => &mut self.ipv4_dialer,
                SocketAddr::V6(_) => &mut self.ipv6_dialer,
            };
            let dialer = match maybe_dialer {
                None => maybe_dialer.insert(Dialer::new::<P>(self.config.clone(), socket_addr)?),
                Some(dialer) => dialer,
            };

            dialer.pending_dials.push_back(to_endpoint);
```

Reply:
Cleaned this code up as part of 0b80573. Thanks for the suggestion!